### DISCUSSION TITLE
How to test sending request.files?

### DISCUSSION THREAD

**Discussion by pamelafox** [pallets/quart#326] [Q&A]
I'm trying to test request.files, and haven't figured out a way to get a fake file to show up in request.files when I'm doing a post() to test_client. I found this gist for Flask:

https://gist.github.com/DazWorrall/1779861

Is that what I should be doing in Quart as well? (I'm surprised it requires a custom request class but that's okay if that's the case)

**Comment by davidism** [comment]
Did passing file data like `client.post("url", data={"avatar": (BytesIO(avatar_bytes), "face.png")})` not work? https://werkzeug.palletsprojects.com/en/3.0.x/test/#request-body (I don't know if Quart is using Werkzeug's test client or an equivalent API yet, but this is how it would be done in Flask.)

I'm not entirely clear what that gist is trying to accomplish, but it was also written 13 years ago before the test client was more developed. The fact that it's still useful to people is a bit alarming.

**Reply by pamelafox** [reply]
I haven't gotten that to work yet, request.files seems to be empty.

**Reply by pamelafox** [reply]
And yes, it seems worth documenting once figured out, given the action on that gist. I grepped hard through the quart codebase to see if there were any examples in the docs/tests.

**Reply by pamelafox** [reply]
I'm glad it's documented in the Flask docs.

**Reply by pamelafox** [reply]
Update: I've discovered that a content-type is needed in Quart for it to get parsed, as is a boundary:

    response = await auth_client.post("/upload",
        headers={"Authorization": "Bearer test", "Content-type": "multipart/form-data; boundary=----WebKitFormBoundarydqGFVE2pINsCjmDQ"},
        data={"file": (BytesIO(b'file contents'), 'hello world.txt')})


However, that's still not right, as it's reporting "No file part in the request". Presumably because I made up that data and it doesn't match each other. I need to read up about form-data boundaries now.

**Reply by pamelafox** [reply]
I'm getting closer but not there yet:


```
    stream = b'------WebKitFormBoundaryr1D8WqBUjhPTDqlM\r\nContent-Disposition: form-data; name="file"; filename="a.txt"\r\nContent-Type: text/plain\r\n\r\n'
    stream += b"foo;bar\n"
    stream += b"\r\n------WebKitFormBoundaryr1D8WqBUjhPTDqlM--\r\n"

    response = await auth_client.post("/upload",
        headers={"Authorization": "Bearer test",
        "Content-type": "multipart/form-data; boundary=----WebKitFormBoundaryr1D8WqBUjhPTDqlM"
        },
        data={"file": (BytesIO(stream), 'helloworld.txt')})
```

**Reply by pamelafox** [reply]
Added content-length:


    stream = b'------WebKitFormBoundaryr1D8WqBUjhPTDqlM\r\nContent-Disposition: form-data; name="file"; filename="a.txt"\r\nContent-Type: text/plain\r\n\r\n'
    stream += b"foo;bar\n"
    stream += b"\r\n------WebKitFormBoundaryr1D8WqBUjhPTDqlM--\r\n"

    response = await auth_client.post("/upload",
        headers={"Authorization": "Bearer test",
        "Content-type": "multipart/form-data; boundary=----WebKitFormBoundaryr1D8WqBUjhPTDqlM",
        "Content-length": str(len(stream))
        },
        data={"file": (BytesIO(stream), 'helloworld.txt')})

According to the debugger, the parser does not see anything being yielded by Request.body. So I'm guessing that I haven't specified the data parameter correctly.

**Reply by pamelafox** [reply]
I figured it out!

```
    stream = b'------WebKitFormBoundaryr1D8WqBUjhPTDqlM\r\nContent-Disposition: form-data; name="file"; filename="a.txt"\r\nContent-Type: text/plain\r\n\r\n'
    stream += b"foo;bar\n"
    stream += b"\r\n------WebKitFormBoundaryr1D8WqBUjhPTDqlM--\r\n"

    response = await auth_client.post(
        "/upload",
        headers={
            "Authorization": "Bearer test",
            "Content-type": "multipart/form-data; boundary=----WebKitFormBoundaryr1D8WqBUjhPTDqlM",
            "Content-length": str(len(stream)),
        },
        files={"file": FileStorage(BytesIO(stream), filename="a.txt")},
    )
```

**Reply by pgjones** [reply]
I think there is something else at play here, as I've written a test in 1a65d32b88a5f462c277ef2b66c3bd97a0529655 and it works as expected. I would also expect Quart's test client to overwrite the Context-Type header when `files` is used. (I would expect the boundary addition to be required for `data` though)

**Reply by pamelafox** [reply]
I'll try mimicing your test exactly- I didn't know about FileStorage in my first attempts, so maybe if I use FileStorage, the other issues go away.

**Reply by pamelafox** [reply]
That worked! So I think the key thing is the usage of the FileStorage data structure. I recommend documenting on https://quart.palletsprojects.com/en/latest/how_to_guides/testing.html and I can send a PR to add it there if you're on board.

**Reply by DanielHabenicht** [reply]
If somebody comes across this and also needs to combine form data with file upload: 
```python
multipartData = {
    "container": "0a5afda2-b745-4b04-9ba0-178ae8662a05",
    "name": f"File-{uuid4()}",
}
files = {
    "data": FileStorage(
        stream=io.BytesIO(b"examplebinary"),
        content_type="application/pdf",
        filename="test.pdf",
    )
}
response = await testclient.post(
    "/documents",
    files=files,
    form=multipartData,
)
```
